src/main/java/core/basesyntax/MyHashMap.java

```diff
@@ -1,19 +1,133 @@
 
     @Override
     public void put(K key, V value) {
+        if (key == null) {
+            putIfKeyIsNull(value);
+            return;
+        }
+        int hash = hashCode(key);
+        int index = indexFor(hash, hashTable.length);
 
+        for (Node<K, V> node = hashTable[index]; node != null; node = node.next) {
+            if (node.hash == hash && (key.equals(node.key))) {
+                node.value = value;
+                return;
+            }
+        }
+        addNode(hash, key, value, index);
     }
 
     @Override
     public V getValue(K key) {
+        if (key == null) {
+            return getForNullKey();
+        }
+
+        int hash = hashCode(key);
+        int index = indexFor(hash, hashTable.length);
+
+        for (Node<K, V> node = hashTable[index]; node != null; node = node.next) {
+            if (node.hash == hash && (key.equals(node.key))) {
+                return node.value;
+            }
+        }
         return null;
     }
 
     @Override
     public int getSize() {
-        return 0;
+        return size;
+    }
+
+    private void putIfKeyIsNull(V value) {
+        Node<K, V> node = hashTable[0];
+        while (node != null) {
+            if (node.key == null) {
+                node.value = value;
+                return;
+            }
+            node = node.next;
+        }
+        addNode(0, null, value, 0);
+    }
+
+    private void addNode(int hash, K key, V value, int bucketIndex) {
+        if (size >= threshold) {
+            resize(2 * hashTable.length);
+            hash = (key != null) ? hashCode(key) : 0;
+            bucketIndex = indexFor(hash, hashTable.length);
+        }
+        Node<K, V> newNode = new Node<>(hash, key, value, hashTable[bucketIndex]);
+        hashTable[bucketIndex] = newNode;
+        size++;
+    }
+
+    private V getForNullKey() {
+        for (Node<K, V> node = hashTable[0]; node != null; node = node.next) {
+            if (node.key == null) {
+                return node.value;
+            }
+        }
+        return null;
+    }
+
+    private void resize(int newCapacity) {
+        Node<K, V>[] newTable = (Node<K, V>[]) new Node[newCapacity];
+        transfer(newTable);
+        hashTable = newTable;
+        threshold = (int) (newCapacity * DEFAULT_LOAD_FACTOR);
+    }
+
+    private void transfer(Node<K, V>[] newTable) {
+        for (Node<K, V> node : hashTable) {
+            while (node != null) {
+                Node<K, V> next = node.next;
+                int index = indexFor(node.hash, newTable.length);
+                node.next = newTable[index];
+                newTable[index] = node;
+                node = next;
+            }
+        }
+    }
+
+    private int indexFor(int hash, int length) {
+        return hash & (length - 1);
+    }
+
+    public int hashCode(K key) {
+        if (key == null) {
+            return 0;
+        }
+        int hash = 17;
+        hash = 31 * hash + key.hashCode();
+        return hash;
```

Comment:
i think we can use this single line instead of all this code
return key == null ? 0 : key.hashCode();